src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs

```diff
@@ -133,59 +171,67 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
                                 severity: NotificationSeverity.Error))
                         {
                             // We handled the command, displayed a notification and did not produce code.
-                            return true;
+                            return;
                         }
+
+                        await TaskScheduler.Default;
                     }
 
                     // reset result
                     result = newResult;
                 }
-                else if (TryNotifyFailureToUser(document, result, waitContext))
+                else if (await TryNotifyFailureToUserAsync(document, result, cancellationToken).ConfigureAwait(false))
                 {
                     // We handled the command, displayed a notification and did not produce code.
-                    return true;
+                    return;
                 }
             }
 
-            var cleanupOptions = document.GetCodeCleanupOptionsAsync(_globalOptions, cancellationToken).AsTask().WaitAndGetResult(cancellationToken);
+            var cleanupOptions = await document.GetCodeCleanupOptionsAsync(_globalOptions, cancellationToken).ConfigureAwait(false);
+            var (formattedDocument, methodNameAtInvocation) = await result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).ConfigureAwait(false);
 
-            // apply the change to buffer
-            // get method name token
-            SyntaxToken methodNameAtInvocation;
-            using (var undoTransaction = _undoManager.GetTextBufferUndoManager(textBuffer).TextBufferUndoHistory.CreateTransaction("Extract Method"))
-            {
-                // apply extract method code to buffer
-                (var formattedDocument, methodNameAtInvocation) = result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).WaitAndGetResult(cancellationToken);
-                formattedDocument.Project.Solution.Workspace.ApplyDocumentChanges(formattedDocument, cancellationToken);
-
-                // apply changes
-                undoTransaction.Complete();
-            }
+            await ApplyChangeAsync(formattedDocument, textBuffer, waitContext).ConfigureAwait(false);
 
             // start inline rename
-            var snapshotAfterFormatting = textBuffer.CurrentSnapshot;
-            var documentAfterFormatting = snapshotAfterFormatting.GetOpenDocumentInCurrentContextWithChanges();
-            _renameService.StartInlineSession(documentAfterFormatting, methodNameAtInvocation.Span, cancellationToken);
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
actually... you raise a good point.  we want the indicator dismissed once we start inline-rename.  so i need to change this logic.

Reply:
oh no.  this is correct.  inline rename will start, and all this code will return and we'll dismiss the BWI.

Reply:
> it makes sense to me, but just wanted to confirm the intention that changes could be applied, then the indicator cancelled before the rename starts.

technically it should not be possible due to there being no opportunity for the user to do anything that would cause the cancellation.  but technically, yes.  if somehow some path canceled that token (again, shouldn't be possible) then we'd apply teh extract method, but not start the inline rename session.